### PR TITLE
v0.9.0

### PR DESCRIPTION
**Pull request by Gusabary**

### CODE REVIEW

File: dock/util/util.hpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    static std::wstring RoomUserStateToString(RoomUser::RoomUserState state) {
+        switch (state) {
+            case RoomUser::ERROR:        return L"Error";
+            case RoomUser::PREPARING:    return L"Preparing";
+            case RoomUser::READY:        return L"Ready";
+            case RoomUser::DISCONNECTED: return L"Disconnected";
+            case RoomUser::PLAYING:      return L"Playing";
+        }
+        throw std::runtime_error("Unsupported RoomUserState.");
+        return {};
```
Comment (Veiasai) [inline dock/util/util.hpp]: could be removed?

File: dock/util/util.hpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    static Element ShowGameSettings(const RoomDetails &details) {
+        if (details.gamename() == "") {
+            return text(L"empty");
+        }
+        if (details.gamename() == "UNO") {
+            return UnoUtil::ShowGameSettings(details.gamesetting());
+        }
+        /// TODO: handle other games
+        throw std::runtime_error("Unsupported game.");
+        return nullptr;
```
Comment (Veiasai) [inline dock/util/util.hpp]: ?

**Review by Veiasai — COMMENTED** [review]
so nice!

**Comment by Veiasai** [comment]
Pls fix the CI,  at least `Linux build and fmt`

File: dock/util/util.hpp (hunk 1 above)
Comment (Gusabary) [reply dock/util/util.hpp]: provide return value on each execution path to suppress compiler warning

File: dock/util/util.hpp (hunk 2 above)
Comment (Gusabary) [reply dock/util/util.hpp]: the same

**Comment by Gusabary** [comment]
> Pls fix the CI, at least `Linux build and fmt`

it will be fixed eventually

File: dock/core/game_attr_fac.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <memory>
+#include <sailgame/common/types.h>
+
+#include "game_attr.h"
+#include "../../games/uno/core/game_attr.h"
```
Comment (Veiasai) [inline dock/core/game_attr_fac.h:7]: couldn't we avoid using '../' ?...

File: dock/core/ui_proxy.h [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+#define RpcMethod_P2(RpcName, p1, p2) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1, p2); \
+    RpcMethodHelper(RpcName);
+
+#define RpcMethod_P3(RpcName, p1, p2, p3) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1, p2, p3); \
+    RpcMethodHelper(RpcName);
+
+#define RpcMethod_P4(RpcName, p1, p2, p3, p4) \
+    auto args = CoreMsgBuilder::Create##RpcName##Args(p1, p2, p3, p4); \
```
Comment (Veiasai) [inline dock/core/ui_proxy.h]: could you try ...args?

Comment (Gusabary) [reply dock/core/ui_proxy.h]: i'm afraid not. the first argument of the macro is method name, with which it's hard to use that variadic template

File: dock/core/game_attr_fac.h (hunk 3 above)
Comment (Gusabary) [reply dock/core/game_attr_fac.h:7]: yep

File: dock/core/ui_proxy.h (hunk 4 above)
Comment (Gusabary) [reply dock/core/ui_proxy.h]: use variadic macros

**Review by Veiasai — APPROVED** [review]

**Pull request merged**